lib/src/layer/polyline_layer.dart

```diff
@@ -39,66 +79,292 @@ class Polyline {
+
+  /// The number of pixels away from the visual line (including any width and
+  /// outline) in which a tap should still register as a tap on the line
+  ///
+  /// Applies to both [onTap] and every [Polyline.onTap].
+  ///
+  /// Defaults to 3.
+  final double onTapTolerance;
+
+  /// Whether a non-tappable [Polyline] should prevent taps from being handled
+  /// on all [Polyline]s beneath it, at overlaps
+  ///
+  /// "Tappable" means that either or both of [Polyline.onTap] and
+  /// [Polyline.tapKey] has been defined.
+  ///
+  /// Applies to both [onTap] and every [Polyline.onTap].
+  ///
+  /// Defaults to `true`.
+  final bool nonTappablesObscure;
+
+  /// Whether a tappable [Polyline] should prevent taps from being handled
+  /// on all [Polyline]s beneath it, at overlaps
+  ///
+  /// "Tappable" means that either or both of [Polyline.onTap] and
+  /// [Polyline.tapKey] has been defined.
+  ///
+  /// If `true`, then [onTap] becomes redundant to [Polyline.onTap].
+  ///
+  /// Defaults to `false`.
+  final bool tappablesObscure;
 
   const PolylineLayer({
     super.key,
     required this.polylines,
-    this.polylineCulling = false,
+    this.onTap,
+    this.onLongPress,
+    this.onSecondaryTap,
+    this.onTapTolerance = 3,
+    this.tappablesObscure = false,
+    this.nonTappablesObscure = true,
+    // TODO: Remove once PR #1704 is merged
+    bool polylineCulling = true,
   });
 
   @override
   Widget build(BuildContext context) {
-    final map = MapCamera.of(context);
+    final camera = MapCamera.of(context);
+
+    final culledPolylines = <Polyline<TapKeyType>>[];
+    bool interactive = onTap != null;
+    for (final line in polylines) {
+      if (!line.boundingBox.isOverlapping(camera.visibleBounds)) continue;
+      if (!interactive && line.onTap != null) interactive = true;
+      culledPolylines.add(line);
+    }
 
-    return MobileLayerTransformer(
-      child: CustomPaint(
-        painter: PolylinePainter(
-          polylineCulling
-              ? polylines
-                  .where((p) => p.boundingBox.isOverlapping(map.visibleBounds))
-                  .toList()
-              : polylines,
-          map,
-        ),
-        size: Size(map.size.x, map.size.y),
-        isComplex: true,
+    final lastHit = _LastHit<TapKeyType>();
+    final paint = CustomPaint(
+      painter: _PolylinePainter<TapKeyType>(
+        polylines: culledPolylines,
+        camera: camera,
+        lastHit: interactive ? lastHit : null,
+        onTapTolerance: onTapTolerance,
+        tappablesObscure: tappablesObscure,
+        nonTappablesObscure: nonTappablesObscure,
       ),
+      size: Size(camera.size.x, camera.size.y),
+      isComplex: true,
+    );
+
+    return MobileLayerTransformer(
+      child: interactive
+          ? MouseRegion(
+              hitTestBehavior: HitTestBehavior.deferToChild,
+              cursor: SystemMouseCursors.click,
+              child: GestureDetector(
```

Comment:
The GestureDetector code has certainly gotten quite involved with all the features, i.e. the cross product between different modal interactions and stacking/obscuring/multi-hit options. I'm slightly worried that this looks a bit brittle...

...but I'm wondering even more if in hindsight and with the desire for so much flexibility, it would have been better to instead pass up the hit detection result so that users could set up a GestureDetector as a parent of the Polyline layer that would leave you with ultimate flexibility w/o the PolylineLayer having to anticipate all use-cases.

Reply:
I know exactly what you mean and I agree it looks a little overgrown, there's suddenly a lot of options. (Although do remember that the in-code docs take up quite a few of the diff lines :D).

However, passing the result up seems less than ideal as well - we're replacing a plugin here so we should be a drop in replacement as far as possible. It would also seem a little half-baked to me.
It would definitely be nice if we could refactor all the handling of all three gesture types, as there's a lot of code duplication there, but I don't see an immediately clear way to do that.
Is there a better middle ground to be struck?

Reply:
> we're replacing a plugin here so we should be a drop in replacement as far as possible.

I still don't get where you got this idea from:

* FlutterMap does not need to achieve feature parity with some third-party plugin nor should it.
* The third party plugin will continue to work, there's no need for anyone to migrate.
* Even despite all your feature parity work, APIs and hit detection semantics are quite different. There's no trivial migration.
* My original PR doesn't share any ancestry with said plugin. Nobody owes anyone anything.

This sounds harsh but I come with good intentions. If you let go of the idea just a little bit,  we could instead discuss what makes the most sense for FM core rather than chasing a predefined goal.

This ties into our earlier discussions, where I wanted to error on the side of adding things incrementally as use-cases arise. You wanted to offer a more feature-full package. Both are fine. The truth is probably somewhere in the middle.

Just keep in mind that there's a real asymmetry: **It will always be easier to add new things, than take things away**

Reply:
I completely understand what you're saying. Me saying "drop in replacement" was a little 'short-sighted'(?).
The only reason I want to give 'credit' is because his plugin is essentially becoming redundant: there's no need for anyone to migrate, but also no need for any new users, and as FM gets away from supported versions, people will migrate away. You're right though that putting it in here isn't appropriate and suggests your work descends from his. Putting it in the docs is a far better idea.

I'll try to drop the idea, let's just leave it here :D.

---

If you want to quickly mockup what you mean by

> but I'm wondering even more if in hindsight and with the desire for so much flexibility, it would have been better to instead pass up the hit detection result so that users could set up a GestureDetector as a parent of the Polyline layer that would leave you with ultimate flexibility w/o the PolylineLayer having to anticipate all use-cases.

, because I don't 100% get what you mean, I'm open to it. Perhaps doing it like that, then providing prebuilt solutions/handlers is the better solution, as we do for some other complex areas (like `TileUpdateTransformer`).

Reply:
> The only reason I want to give 'credit' is because his plugin is essentially becoming redundant: there's no need for anyone to migrate, but also no need for any new users, and as FM gets away from supported versions, people will migrate away. 

My point is, that that's only true *because* you're striving for feature parity. There's tons of 3rd party implementations of layers that have overlapping functionality with core features. There's marker layers with clustering, marker layers with popups, tile providers with offline storage, ... . None of them are redundant, nor should the FlutterMap core try to replicate full set of functionality.
By going into this review making some 3rd party plugin the bottom line we're painting ourselves into a corner. Supporting only a core subset of functionality is ok, likely even desirable. Folks can always ask us to add more features, they can always continue to use someones plugin, or build their own.

The documentation also **should not** acknowledge that there "once was a plugin that did something similar". The docs aren't a fairy tale or archeological piece. Nobody who reads the docs is looking for that information. What the docs **could or should** acknowledge is that there are these curated and working plugins with similar or extended feature sets. If a plugin is fringe, poor quality or defunct, it shouldn't be mentioned.

Look, I'm merely trying to pivot the conversation to: "These are great common use-cases we want to support from the get go. And these are great common use-cases we might want to support in the future" instead of "We need to do X because plugin Y does it".

> , because I don't 100% get what you mean, I'm open to it.

Known problem, most of the time I've no clue what I mean :)

There's the famous quote of "all problems can be solved by another layer of abstraction, except the problem of too many abstractions". Sometimes folks call for "deep modules", which just means any layer should abstract "some good" amount of functionality in order to justify its existence. Otherwise it might as well go away.

Looking at our concrete case,  I see what I would call a "thin abstraction", i.e. the abstraction the PolylineLayer is mostly wrapping GestureDetector/MouseRegion functionality through w/o providing much of an abstraction. Layering this abstractions this way also opens up the rabbit hole of wrapping more and more functionality through (there's still tons of interaction modes missing). For example, my own fork supports mouse hovers to change the color of polylines when the pointer goes over them :). This is not a call for us to wrap even more features. I deliberately ripped it out for this PR because I didn't think it was as generally useful and if there's demand we can always add in the future. Meanwhile I'll maintain it for my use-case.

An essay later, what I was proposing is to change the layering, **to push responsibilities up the stack** giving users all the flexibility w/o PolygonLayer having to support all the fringe use-cases. Imagine an API that would look more like this:

```dart
Widget build(context) {
  final lastHit = LastHit();

  void setHit(LastHit hit) => lastHit = hit;

  return GestureDetector(
    hitTestBehavior: HitTestBehavior.deferToChild,
    onTap: () {
       if (lastHit.hit == null) return;
      
       // Do something.
    },
    child: PolygonLayer(
      polylines: [ /* ... */ ],
      hitSetter: setHit,
    ),
  }
}
```

You're probably going to say that this is harder to use and you're right. We could still support simple taps internally while offering this mechanism for boon like me to get fringe hover functionality w/o having to explicitly support it... at least until there's consensus that everyone wants hover.

Reply:
I tend to agree with @ignatz here, this implementation aims to offer a 1st party alternative to `flutter_map_tappable_polyline`, we don't have to provide exactly the same features nor API. It's not like the package will be deleted after this package is merged, it might become less relevant in the long term but migrating will be the responsibility of the developers who want to migrate, not ours.



> If you want to quickly mockup what you mean by
> 
> > but I'm wondering even more if in hindsight and with the desire for so much flexibility, it would have been better to instead pass up the hit detection result so that users could set up a GestureDetector as a parent of the Polyline layer that would leave you with ultimate flexibility w/o the PolylineLayer having to anticipate all use-cases.
> 
> , because I don't 100% get what you mean, I'm open to it.

I think what he meant here would be that to have a fully flexible implementation of tap events we would need to ensure that tap events emanating from the `PolylineLayer` would propagate correctly so, a `GestureDetector` would be able to catch those events by itself without needing us to add more callbacks in the API to handle some specific use cases.

Reply:
That does seem like a sensible middle ground. We can just send all the hit polylines up, then the user can decide which are tappable, and which are not, and whether to handle them all, or just the top. I agree that that is probably a better solution.
(I wonder if there is a way to eliminate the `lastHit` stuff from being public, as that is less than ideal.)

Reply:
Agreed, on `lastHit` being a bit ugly, I was mostly trying to keep it similar in my mock. 

A more idiomatic way could look like:

```dart
// Could be a class or record, don't really care
class Hit {
 List<Polylines> lines,
 LatLng point,
};

// ...

final notifier = ValueNotifier<Hit?>(null);
return PolylineLayer(
  //...
  hitNotifier: notifier,
);
``` 

With this approach you could also listen to hits.

Reply:
@ignatz Yeah I like that. How does the implementation look now?

Reply:
@ignatz Yeah I like that. How does the implementation look now?